draft-ietf-jsonpath-base.md

```diff
@@ -1104,6 +1094,23 @@ Some examples:
 | `[3,4,5]` | `$[?index(@)==2]`<br>`$[?index(@)==17]` | `[5]`<br>`[]` | Select array element |
 | `{"a":{"b":{5},c:0}}` | `$[?@.b==5 && !@.c]` | `[{"b":{5},c:0}]` | Existence  |
 
+
+## Normalized Paths
+
+A Normalized Path identifies a node by providing a query that results in exactly that node. For example,
+the JSONPath expression `$.book[?(@.price<10)]` could select two values with Normalized Paths
+`$.book[3]` and `$.book[5]`.
+
```

Comment:
`$['book'][3]` and `$["book"][3]` address the same node, so the Normalized Path is canonical in structure only, not in syntax.

Reply:
No ... even here we should state in the spec, that `$['book'][3]` is the one and only Normalized Path. Reason here is to reuse normalized pathes as strings in JSON documents without necessity of escaping ... exactly the same reason, why we support `'...'` additionally  to JSON's `"..."`.

Reply:
I like the idea of defining the preferred (canonical) syntax.  We still need to escape backslashes and decide how certain characters are shown (e.g., \r vs. \u000d, no \/ etc.) -- should not be too hard, but needs to be done.

Reply:
I'm not sure of the goal here, or what the WG consensus is likely to be, especially since Normalized Paths don't currently feature much in the spec. Should we take this discussion to the mailing list?

Reply:
I don't know, what the goal of a mailing list discussion would be.

IIRC normalized pathes as an alternative result (to values) of queries never have been questionable. It is also the only way to verify, that identical queries lead to identical results ... wouldn't be possible with ambiguous node values only ... it's key for tests!

But if you want ... take this discussion to the mailing list?

Reply:
> I think this is a tangent, but a useful one. Should there be a single canonical form for a Normalized Path?

Yes, that's the question. But not everyone who can usefully contribute will be tuned in to this PR.

> 
> (The discussion about "Normalized Path" vs. "Output Path" is a bit confused, like a discussion whether somebody is an "MIT Alumnus" or "CEO of acme corp." -- these are just orthogonal, one is what it is, one is where it is used.)

I understand, but I don't see the need for "Output Path" as a separate term. We can simply talk about the output from an implementation etc.

Reply:
> I understand, but I don't see the need for "Output Path" as a separate term. We can simply talk about the output from an implementation etc.

I don't see a need to *introduce* this term either, but we can say that the output of a query can be values and/or paths, and these are then of course output values and output paths without further definition or capitalization.

Let's open the Normalized Path determinism ("canonicalization") issue on the mailing list, though; we can finish the rest of this PR while this is being discussed.

Reply:
'Output Path' is not needed. Result of a JSONPath query is a list of nodes formatted as canonical Normalized Pathes.

Reply:
@cabo @goessner "canonical Normalized Path" is a mouthful. Why not simply "Canonical Path"?

Reply:
Normalized Path.

(Don't use the C word.)

Reply:
For the record, what's wrong with "canonical"?

Reply:
Generally evokes enormous amounts of pain in people who have ever tried to do c14n.

Also, unnecessary here, as we already use the term "normalized".